### PR TITLE
[flang] Remove hardcoded bits from AddDebugInfo.

### PR DESCRIPTION
**Pull request by abidh**
This PR adds following options to the AddDebugInfo pass.

1. IsOptimized flag.
2. Level of debug info to generate.
3. Name of the source file

This enables us to remove the hard coded values from the code. It also allows us to test the pass with different options. The tests have been modified to take advantage of that.

The calling convention flag and producer name have also been improved.

### CODE REVIEW

File: flang/include/flang/Tools/CLOptions.inc [hunk 1]
```diff
@@ -283,34 +301,35 @@ inline void createOpenMPFIRPassPipeline(
...
     return;
   case llvm::codegenoptions::NoDebugInfo:
     return;
   default:
     // TODO: Add cases and passes for other debug options.
     // All other debug options not implemented yet, currently emits warning
     // and generates as much debug information as possible.
-    addDebugInfoPass(pm);
+    addDebugInfoPass(pm, debugLevel, OptLevel, inputFilename);
     return;
```
Comment (jeanPerier) [inline flang/include/flang/Tools/CLOptions.inc:319]: Given you are forwarding the debug level to the pass now, it seems like you can transform the switch into:
```
if (debugLevel != llvm::codegenoptions::NoDebugInfo)
  addDebugInfoPass(pm, debugLevel, OptLevel, inputFilename);
```

File: flang/lib/Optimizer/Transforms/AddDebugInfo.cpp [hunk 2]
```diff
@@ -52,21 +55,40 @@ void AddDebugInfoPass::runOnOperation() {
...
+  if (inputFilename.empty()) {
+    if (auto fileLoc = module.getLoc().dyn_cast<mlir::FileLineColLoc>()) {
+      fileName = llvm::sys::path::filename(fileLoc.getFilename().getValue());
+      filePath = llvm::sys::path::parent_path(fileLoc.getFilename().getValue());
+    } else
+      fileName = "-";
+  } else {
+    fileName = inputFilename;
+    llvm::SmallString<256> cwd;
+    if (!llvm::sys::fs::current_path(cwd))
```
Comment (jeanPerier) [inline flang/lib/Optimizer/Transforms/AddDebugInfo.cpp:80]: Isn't `current_path` the working directory where the compiler is being run? If so, that could be different from the file directory.

File: flang/lib/Optimizer/Transforms/AddDebugInfo.cpp [hunk 3]
```diff
@@ -52,21 +55,40 @@ void AddDebugInfoPass::runOnOperation() {
...
+  // 1. Name of the file as was presented to compiler. This can be absolute
+  // or relative to 2.
+  // 2. Current working directory
+  //
+  // We are also dealing with 2 different situations below. One is normal
+  // compilation where we will have a value in 'inputFilename' and we can
+  // obtain the current directory using 'current_path'.
+  // The 2nd case is when this pass is invoked directly from 'fir-opt' tool.
+  // In that case, 'inputFilename' may be empty. Location embedded in the
+  // module will be used to get file name and its directory.
```
Comment (jeanPerier) [inline flang/lib/Optimizer/Transforms/AddDebugInfo.cpp:70]: Why not always using the module location since it must have been provided?

File: flang/lib/Optimizer/Transforms/AddDebugInfo.cpp [hunk 4]
```diff
@@ -75,43 +97,46 @@ void AddDebugInfoPass::runOnOperation() {
...
         mlir::StringAttr::get(context, funcOp.getName());
     mlir::LLVM::DIBasicTypeAttr bT = mlir::LLVM::DIBasicTypeAttr::get(
         context, llvm::dwarf::DW_TAG_base_type, "void", /*sizeInBits=*/0,
         /*encoding=*/1);
     mlir::LLVM::DISubroutineTypeAttr subTypeAttr =
-        mlir::LLVM::DISubroutineTypeAttr::get(
-            context, llvm::dwarf::getCallingConvention("DW_CC_normal"),
-            {bT, bT});
-    mlir::LLVM::DIFileAttr funcFileAttr = getFileAttr(funcFilePath);
+        mlir::LLVM::DISubroutineTypeAttr::get(context, CC, {bT, bT});
```
Comment (jeanPerier) [inline flang/lib/Optimizer/Transforms/AddDebugInfo.cpp:115]: The subroutine type is still incorrect (always set to void(void)), right? If so, that is OK, but could you add a FIXME comment so that it is clear this is temporary.

File: flang/lib/Optimizer/Transforms/AddDebugInfo.cpp [hunk 5]
```diff
@@ -75,43 +97,46 @@ void AddDebugInfoPass::runOnOperation() {
...
+    mlir::LLVM::DISubprogramFlags subprogramFlags =
+        mlir::LLVM::DISubprogramFlags{};
+    if (isOptimized)
+      subprogramFlags = mlir::LLVM::DISubprogramFlags::Optimized;
     if (!funcOp.isExternal()) {
       id = mlir::DistinctAttr::create(mlir::UnitAttr::get(context));
       compilationUnit = cuAttr;
-      subprogramFlags =
-          subprogramFlags | mlir::LLVM::DISubprogramFlags::Definition;
+      subprogramFlags = mlir::LLVM::DISubprogramFlags::Definition;
```
Comment (jeanPerier) [inline flang/lib/Optimizer/Transforms/AddDebugInfo.cpp:129]: Why this change? Isn't it overriding the Optimized flag if it was set?

File: flang/lib/Optimizer/Transforms/AddDebugInfo.cpp [hunk 6]
```diff
@@ -75,43 +97,46 @@ void AddDebugInfoPass::runOnOperation() {
...
     }
     auto spAttr = mlir::LLVM::DISubprogramAttr::get(
         context, id, compilationUnit, fileAttr, funcName, funcName,
-        funcFileAttr,
-        /*line=*/1,
-        /*scopeline=*/1, subprogramFlags, subTypeAttr);
+        funcFileAttr, /*line=*/1, /*scopeline=*/1, subprogramFlags,
```
Comment (jeanPerier) [inline flang/lib/Optimizer/Transforms/AddDebugInfo.cpp:133]: line and scopeline are still bogus here (should come from the funcOp location), right? Could you add a FIXME?

**Review by jeanPerier — COMMENTED** [review]
Thanks! My main comment is that I do not get why you need to pass the filename from the driver, it was not hardcoded since it came from the module location which is created in lowering from the source file name.

**Comment by abidh** [comment]
> Thanks! My main comment is that I do not get why you need to pass the filename from the driver, it was not hardcoded since it came from the module location which is created in lowering from the source file name.

The name and path field in the DICompileUnit are used to set the `DW_AT_name` and `DW_AT-comp_dir` attribute in the `DWARF`. Here is the description of these 2 attributes from the DWARF5 standard.

```
A DW_AT_name attribute whose value is a null-terminated string containing
the full or relative path name (relative to the value of the DW_AT_comp_dir
attribute, see below) of the primary source file from which the compilation
unit was derived.

A DW_AT_comp_dir attribute whose value is a null-terminated string
containing the current working directory of the compilation command that
produced this compilation unit in whatever form makes sense for the host
system 
```

Please also see the example below. It demonstrates where the gfortran was invoked from and what we get in the `DWARF`. Passing file from driver to AddDebugInfo will allow us to generate similar `DWARF
```
/home/haqadeer/build$ gfortran ../src/test.f90
              DW_AT_name	("../src/test.f90")
              DW_AT_comp_dir	("/home/haqadeer/build")
```

File: flang/lib/Optimizer/Transforms/AddDebugInfo.cpp (hunk 2 above)
Comment (abidh) [reply flang/lib/Optimizer/Transforms/AddDebugInfo.cpp:80]: As I explained in my earlier comment, `DWARF` requires working directory.

File: flang/lib/Optimizer/Transforms/AddDebugInfo.cpp (hunk 3 above)
Comment (abidh) [reply flang/lib/Optimizer/Transforms/AddDebugInfo.cpp:70]: In ideal case, we would like it to be the current working directory as I explained above. But we don't have that information in 2nd case so using parent directory is sort of a compromise.

File: flang/lib/Optimizer/Transforms/AddDebugInfo.cpp (hunk 4 above)
Comment (abidh) [reply flang/lib/Optimizer/Transforms/AddDebugInfo.cpp:115]: I will do that.

File: flang/lib/Optimizer/Transforms/AddDebugInfo.cpp (hunk 5 above)
Comment (abidh) [reply flang/lib/Optimizer/Transforms/AddDebugInfo.cpp:129]: Thanks for catching this. My initial working version did not set the optimizated flag outside of if so that I removed the OR. I will remove this change from the PR.

File: flang/lib/Optimizer/Transforms/AddDebugInfo.cpp (hunk 6 above)
Comment (abidh) [reply flang/lib/Optimizer/Transforms/AddDebugInfo.cpp:133]: I will do that.

File: flang/include/flang/Tools/CLOptions.inc (hunk 1 above)
Comment (abidh) [reply flang/include/flang/Tools/CLOptions.inc]: Thanks for the suggestion. Done now.

File: flang/lib/Optimizer/Transforms/AddDebugInfo.cpp [hunk 7]
```diff
@@ -52,21 +55,40 @@ void AddDebugInfoPass::runOnOperation() {
   mlir::ModuleOp module = getOperation();
   mlir::MLIRContext *context = &getContext();
   mlir::OpBuilder builder(context);
-  std::string inputFilePath("-");
-  if (auto fileLoc = module.getLoc().dyn_cast<mlir::FileLineColLoc>())
-    inputFilePath = fileLoc.getFilename().getValue();
+  llvm::StringRef fileName;
+  std::string filePath;
```
Comment (tblah) [inline flang/lib/Optimizer/Transforms/AddDebugInfo.cpp:59]: Does DWARF support paths with wide characters?

Comment (abidh) [reply flang/lib/Optimizer/Transforms/AddDebugInfo.cpp:59]: `DWARF` has `DW_AT_use_UTF8` attribute which can be used to say that program entities and filenames use `UTF-8` representation.

**Review by jeanPerier — APPROVED** [review]
Thanks for explaining. I understand now that the pervious debug info was incorrect with regard to the build directory:

````
/home/haqadeer/build$ gfortran ../src/test.f90
              DW_AT_name	("test.f90")
              DW_AT_comp_dir	("/home/src")
````

Too bad there is no equivalent of std::filesystem::relative in llvm filesystem, this would have allowed to easily compute the inputFileName from the mlir location and current working directory without an extra pass option (there is not usages of std::filesystem in LLVM outside of libcxx and some tests, so probably best to not introduce it just for this).

**Comment by vzakhari** [comment]
Thank you for this work @abidh! Do you know or have some pointers to how Clang handles `DW_AT_name` and `DW_AT_comp_dir` generation in the LTO case, when multiple LLVM modules end up being linked into a single LLVM module? Is it using some artificial module name in this case?

File: flang/lib/Optimizer/Transforms/AddDebugInfo.cpp (hunk 7 above)
Comment (tblah) [reply flang/lib/Optimizer/Transforms/AddDebugInfo.cpp:59]: I see that some of the functions in `llvm::sys::fs` represent paths as `Twine`s so I guess a `std::string` should be good enough for now. Off the top of my head, I can't think of a reason why a `std::string` can't store a UTF8 string (::size() might be confusing, but that shouldn't matter here).

**Review by tblah — APPROVED** [review]

**Comment by abidh** [comment]
> Thank you for this work @abidh! Do you know or have some pointers to how Clang handles `DW_AT_name` and `DW_AT_comp_dir` generation in the LTO case, when multiple LLVM modules end up being linked into a single LLVM module? Is it using some artificial module name in this case?

I have seen that `gcc` can generate compilation unit with `<artifical>` as file name when using lto. Please see discussion in the following thread.
https://lists.dwarfstd.org/pipermail/dwarf-discuss/2022-November/002180.html

I am not sure whether `clang` generates an artificial compile unit or not. I tried a simple example with 2 files with both `clang` and `gcc`. I see that `clang` generates 2 compile units with actual file name. `GCC` generates one extra compile unit with `<artificial>` name. The `DW_AT_comp_dir` is same in all case and point to the current working directory.
```

$ clang -flto -g  a.c main.c
DW_AT_name        : (indexed string: 0x1): a.c
DW_AT_name        : (indexed string: 0x5): main.c

$ gcc -flto -g  a.c main.c
DW_AT_name        : (indirect line string, offset: 0x0): <artificial>
DW_AT_name        : (indirect line string, offset: 0x2e): a.c
DW_AT_name        : (indirect line string, offset: 0x32): main.c

```
This probably does not answer your question. I will try to dig more.
